packages/workflow/src/utils.ts

```diff
@@ -1,12 +1,15 @@
 /* eslint-disable @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-argument */
-export const deepCopy = <T>(source: T): T => {
+export const deepCopy = <T>(source: T, hash = new WeakMap()): T => {
 	let clone: any;
 	let i: any;
 	const hasOwnProp = Object.prototype.hasOwnProperty.bind(source);
-	// Primitives & Null
-	if (typeof source !== 'object' || source === null) {
+	// Primitives & Null & Function
+	if (typeof source !== 'object' || source === null || source instanceof Function) {
 		return source;
 	}
+	if (hash.has(source)) {
+		return hash.get(source);
```

Comment:
Should we use WeakSet instead, and return `[Circular Dependency]` string in those cases.

Reply:
It depends on the use case. There can be scenarios when you use a property path and expect some value and not `[Circular Dependency]`, in that case you would have to have a guard function that checks if the requested prop is a circular dep

Reply:
For reference MDN [uses WeakSet](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Cyclic_object_value#circular_references) for the `JSON.stringify()` replacer arg.

Reply:
Doesn't really matter which we're using the point is returning the value or nothing

Reply:
If we return the value we need to use `WeakMap` if we just exit the loop iteration (and thus returning `undefined`) we can use `WeakSet`